1-Draft/RFCNNNN-RFC-Native-Command-Exit-Errors.md

```diff
@@ -0,0 +1,206 @@
+---
+RFC: RFCNNNN-RFC-Native-Command-Exit-Errors
+Author: Micheal Padden
+Status: Draft
+Area: Engine, Language
+Version: 1.0.0
+---
+
+# Native Command Error Handling
+
+Although Powershell has an exception-based error handling framework, it
+does not currently apply to native commands. 
+
+Exception-based error handling makes it easier to write robust code
+as less boilerplate code is needed to check and handle errors.
+
+Powershell scripts using native commands would benefit from being able
+to use error handling features like those used by cmdlets.
+
+## Motivation
+
+Native commands return an exit code to the calling application which
+will be zero for success or non-zero for failure. A robust script will
+not assume success, instead checking the exit code after any statement
+calling a native command, using boilerplate like the following:
+```Powershell
+ if( ! $? ) 
+ { 
+     exit $lastexitcode; 
+ }
+```
+
+The Bourne again shell provides a `set -eo pipefail` exit code handling mode
```

Comment:
My understanding of `-o pipefail` is that it does two things:

1. stop in the middle of a pipeline if any native exe has non-zero exit
2. preserve exit code so subsequent commands with zero exit code doesn't clobber it

Current PowerShell behavior has subsequent native commands clobber `$LastExitCode` today.  So I would suggest we don't make any change here.  We are just adopting the first behavior above which is if any native exe has non-zero exit and `$PSNativeCommandError` = `Stop`, then it terminates right away.

Reply:
@SteveL-MSFT: `set -o pipefail` doesn't impact the _control flow_: the pipeline completes as it normally would; the only difference is that the first command participating in the pipeline that reports a non-zero exit code, if any, determines the pipeline's overall exit code; otherwise, it is the _last_ command (which otherwise _alone_ determines the exit code).

The only way for a pipeline in a POSIX-like shell to terminate prematurely is for a participating command to stop reading from the pipeline prematurely, as `head` does, for instance (by design, and it therefore itself reports exit code `0`).

Otherwise, the pipeline runs to completion, and the `-e` option, if set, then acts on whatever the pipeline's overall exit code is.